crates/turbopack-ecmascript/src/swc_comments.rs

```diff
@@ -0,0 +1,122 @@
+use std::{collections::HashMap, mem::take};
+
+use swc_core::{
+    base::SwcComments,
+    common::{
+        comments::{Comment, Comments},
+        BytePos,
+    },
+};
+
+/// Immutable version of [SwcComments] which doesn't allow mutation. The `take`
+/// variants are still implemented, but do not mutate the content. They are used
+/// by the SWC Emitter.
+pub struct ImmutableComments {
+    pub leading: HashMap<BytePos, Vec<Comment>>,
+    pub trailing: HashMap<BytePos, Vec<Comment>>,
+}
+
+impl ImmutableComments {
+    pub fn new(comments: SwcComments) -> Self {
+        Self {
+            leading: comments
+                .leading
+                .iter_mut()
+                .filter_map(|mut r| {
+                    let c = take(r.value_mut());
+                    (!c.is_empty()).then_some((*r.key(), c))
+                })
+                .collect(),
+            trailing: comments
+                .trailing
+                .iter_mut()
+                .filter_map(|mut r| {
+                    let c = take(r.value_mut());
+                    (!c.is_empty()).then_some((*r.key(), c))
+                })
+                .collect(),
+        }
+    }
+}
+
+impl Comments for ImmutableComments {
```

Comment:
I think this should implement `with_leading`.

Reply:
I don't remember if codegen uses `with_leading`, but some transforms uses it.

https://github.com/swc-project/swc/blob/662f236aa5c1375ea4b175dbb6d85ad974503b84/crates/swc_ecma_transforms_typescript/src/strip.rs#L313-L315

Reply:
And well, `with_leading` calls `take_leading` and `add_leading_comments` by default

Reply:
Transformation code will receive a mutable [SwcComments](https://github.com/vercel/turbo/blob/f068a992b79a6abc4b41308f13ecfddba29cb93f/crates/turbopack-ecmascript/src/parse.rs#L246). It's only transformed into this immutable comments after transformation, when we insert the `ParseResult::Ok` into our cached graph.

Reply:
implemented them. Seems like they are unused currently, but it doesn't hurt to have non-panicing variants for them.